src/base/tristatebool.h

```diff
@@ -38,27 +38,32 @@ class TriStateBool
 
     constexpr TriStateBool() = default;
     constexpr TriStateBool(const TriStateBool &other) = default;
-    explicit constexpr TriStateBool(int value)
-        : m_value(value < 0 ? -1 : (value > 0 ? 1 : 0))
+    explicit constexpr TriStateBool(const bool boolean)
     {
+        *this = boolean ? True : False;
     }
 
-    explicit constexpr operator int() const
+    TriStateBool &operator=(const TriStateBool &other) = default;  // add constexpr when using C++17
+
+    explicit constexpr operator signed char() const
     {
         return m_value;
     }
 
-    TriStateBool &operator=(const TriStateBool &other) = default;  // add constexpr when using C++17
-
-    constexpr friend bool operator==(const TriStateBool &left, const TriStateBool &right)
+private:
+    explicit constexpr TriStateBool(const int value)
+        : m_value((value < 0) ? -1 : ((value > 0) ? 1 : 0))
     {
-        return (left.m_value == right.m_value);
     }
 
-private:
     signed char m_value = -1; // Undefined by default
 };
 
+constexpr bool operator==(const TriStateBool &left, const TriStateBool &right)
+{
+    return (static_cast<signed char>(left) == static_cast<signed char>(right));
```

Comment:
IMO, it's bad idea to use type casting here. It's better to make it friend and compare underlying value directly.

Reply:
Oops... it was correct before.

Reply:
>IMO, it's bad idea to use type casting here. It's better to make it friend and compare underlying value directly.

You sure even when the casting is to the same internal type?

Reply:
Yes. Type comparison is its internal logic.
At the very least, it's suspicious that in order to compare values of some type, we have to convert them to another type.

Reply:
reverted, although it is acceptable for me if the conversion done can represent the salient state of the type and guarantee the correctness.